### Source : https://github.com/adeshpande3/Pandas-Tutorial

# Introduction

In [1]:
import numpy as np
import pandas as pd

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use. 

Just think of it as a table for now. 

In [2]:
df = pd.read_csv('RegularSeasonCompactResults.csv')

In [9]:
hank = pd.read_clipboard(header=None,sep=",")

In [10]:
hank

0   1   2   3    4   5   6   7   8   9   10  11  12  13  14  15  16  17
0   1  13  21  11  196  75   4   3  34   6   7   8   0   1   2   3   4   5
1   3  42  12  33  766  75   4  55   6   4   3   4   5   6   7   0  11  12
2   1  22  33  11  999  11   2   1  78   0   1   2   9   8   7   1  76  88

In [13]:
# dfjson = pd.read_json("Files/sample.json")

In [14]:
df.to_csv('example.csv', index=False)

In [18]:
df.to_excel('example.xlsx', index=False)

In [19]:
df.to_csv('example.txt', index=False)

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [21]:
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

In [32]:
type(df)

pandas.core.frame.DataFrame

In [20]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [22]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the the **shape** attribute

In [23]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [24]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [34]:
df.describe()

Season         Daynum          Wteam         Wscore  \
count  145289.000000  145289.000000  145289.000000  145289.000000   
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [31]:
df.max().values

array([2016, 132, 1464, 186, 1464, 150, 'N', 6], dtype=object)

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [35]:
df['Wscore']

0         81
1         77
2         63
3         70
4         86
          ..
145284    70
145285    72
145286    82
145287    66
145288    87
Name: Wscore, Length: 145289, dtype: int64

In [36]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score. 

In [37]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [38]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [39]:
df['Wscore'][24970]

186

In [40]:
df['Season'].value_counts()

2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: Season, dtype: int64

# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [41]:
df['Wscore'].argmax()

24970

In [42]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [43]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [63]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

In [45]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [46]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [47]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [48]:
df.iloc[5:9]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5    1985      25   1218      79   1337      78    H      0
6    1985      25   1228      64   1226      44    N      0
7    1985      25   1242      58   1268      56    N      0
8    1985      25   1260      98   1133      80    H      0

In [69]:
df.iloc[df['Wscore'].argmax(),[0,3]]

Season    1991
Wscore     186
Name: 24970, dtype: object

In [50]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [64]:
df.loc[df['Wscore'].argmax(), ['Lscore','Season']]

Lscore     140
Season    1991
Name: 24970, dtype: object

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [52]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [70]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
100027    2008      66   1203      49   1387      20    H      0
49310     1997      66   1157      61   1204      21    H      0
89021     2006      44   1284      41   1343      21    A      0
85042     2005      66   1131      73   1216      22    H      0
103660    2009      26   1326      59   1359      22    H      0

In [72]:
df.groupby('Lscore').mean()

Season  Daynum   Wteam      Wscore        Lteam     Numot
Lscore                                                           
20      2008.0    66.0  1203.0   49.000000  1387.000000  0.000000
21      2001.5    55.0  1220.5   51.000000  1273.500000  0.000000
22      2007.0    46.0  1228.5   66.000000  1287.500000  0.000000
23      2004.0    54.0  1214.0   66.333333  1253.333333  0.000000
24      2003.5    23.5  1294.5   81.000000  1213.000000  0.000000
...        ...     ...     ...         ...          ...       ...
137     1990.0    38.0  1258.0  152.000000  1109.000000  0.000000
140     1993.0    81.0  1355.0  156.666667  1264.000000  0.666667
141     1990.0    96.0  1261.0  148.000000  1258.000000  0.000000
144     1989.0    68.0  1258.0  162.000000  1109.000000  0.000000
150     1989.0    92.0  1258.0  181.000000  1109.000000  0.000000

[117 rows x 6 columns]

In [74]:
df.groupby('Season').median()

Daynum   Wteam  Wscore   Lteam  Lscore  Numot
Season                                               
1985      84.0  1281.0    74.0  1279.0    64.0    0.0
1986      82.0  1281.0    74.0  1279.0    63.0    0.0
1987      84.0  1283.0    77.0  1278.0    66.0    0.0
1988      82.0  1281.0    79.0  1280.0    67.0    0.0
1989      82.0  1281.0    81.0  1281.0    69.0    0.0
1990      82.0  1278.0    80.0  1284.0    68.0    0.0
1991      82.0  1286.0    82.0  1278.0    69.0    0.0
1992      79.0  1281.0    79.0  1283.0    67.0    0.0
1993      82.0  1281.0    79.0  1284.0    67.0    0.0
1994      82.0  1285.0    80.0  1283.0    68.0    0.0
1995      82.0  1290.0    79.0  1278.0    67.0    0.0
1996      82.0  1283.0    77.0  1282.0    65.0    0.0
1997      82.0  1283.0    76.0  1281.0    63.0    0.0
1998      81.0  1286.0    76.0  1282.0    64.0    0.0
1999      80.0  1283.0    75.0  1281.0    63.0    0.0
2000      77.0  1284.0    76.0  1278.0    63.0    0.0
2001      78.0  1280.0    77.0  1282.0    64.0    0.0
2002      78.0  1282.0    76.0  1281.0    64.0    0.0
2003      75.0  1281.0    75.0  1280.0    63.0    0.0
2004      77.0  1283.0    75.0  1276.0    63.0    0.0
2005      77.0  1283.0    74.0  1279.0    62.0    0.0
2006      77.0  1279.0    74.0  1286.0    62.0    0.0
2007      75.0  1283.0    74.0  1283.0    63.0    0.0
2008      72.0  1288.0    74.0  1280.0    62.0    0.0
2009      73.0  1288.0    73.0  1278.0    62.0    0.0
2010      73.0  1291.0    74.0  1276.0    62.0    0.0
2011      73.0  1284.0    74.0  1280.5    62.0    0.0
2012      73.0  1283.0    73.0  1284.0    61.0    0.0
2013      72.0  1282.0    72.0  1287.0    61.0    0.0
2014      73.0  1288.0    75.0  1279.0    64.0    0.0
2015      73.0  1291.0    72.0  1279.0    61.0    0.0
2016      73.0  1290.0    78.0  1278.0    66.0    0.0

In [ ]:
df

# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [75]:
df[df['Wscore'] > 150]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5269     1986      75   1258     151   1109     107    H      0
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
16040    1989      40   1328     152   1331     122    H      0
16853    1989      68   1258     162   1109     144    A      0
17867    1989      92   1258     181   1109     150    H      0
19653    1990      30   1328     173   1109     101    H      0
19971    1990      38   1258     152   1109     137    A      0
20022    1990      40   1116     166   1109     101    H      0
22145    1990      97   1258     157   1362     115    H      0
23582    1991      26   1318     152   1258     123    N      0
24341    1991      47   1328     172   1258     112    H      0
24970    1991      68   1258     186   1109     140    H      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
40060    1995      32   1375     156   1341     114    H      0
52600    1998      33   1395     153   1410      87    H      0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [76]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
52600    1998      33   1395     153   1410      87    H      0

# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [90]:
df.groupby('Wteam')

In [95]:
df.groupby('Wteam').mean()

Season     Daynum     Wscore        Lteam     Lscore     Numot
Wteam                                                                     
1101   2015.388889  75.666667  78.111111  1241.277778  69.222222  0.000000
1102   2003.317152  68.126214  69.893204  1296.711974  57.838188  0.045307
1103   2002.482625  76.959459  75.839768  1268.266409  63.189189  0.057915
1104   1999.932677  72.535304  75.825944  1278.689655  62.425287  0.031199
1105   2007.111732  89.720670  74.960894  1252.726257  64.318436  0.055866
...            ...        ...        ...          ...        ...       ...
1460   2002.717949  74.622378  75.531469  1291.808858  64.111888  0.030303
1461   2000.227459  71.237705  75.170082  1265.122951  63.770492  0.051230
1462   2001.055800  76.157122  79.906021  1259.917768  65.198238  0.030837
1463   2001.585242  78.539440  71.720102  1222.860051  61.521628  0.045802
1464   2000.968310  72.982394  73.926056  1286.352113  64.524648  0.038732

[364 rows x 6 columns]

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [78]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       A         3
       N         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: Wloc, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [80]:
ar = df.values

Now, you can simply just access elements like you would in an array. 

In [84]:
dfnew = pd.DataFrame(ar,columns=df.columns.tolist())

In [85]:
dfnew

Season Daynum Wteam Wscore Lteam Lscore Wloc Numot
0        1985     20  1228     81  1328     64    N     0
1        1985     25  1106     77  1354     70    H     0
2        1985     25  1112     63  1223     56    H     0
3        1985     25  1165     70  1432     54    H     0
4        1985     25  1192     86  1447     74    H     0
...       ...    ...   ...    ...   ...    ...  ...   ...
145284   2016    132  1114     70  1419     50    N     0
145285   2016    132  1163     72  1272     58    N     0
145286   2016    132  1246     82  1401     77    N     1
145287   2016    132  1277     66  1345     62    N     0
145288   2016    132  1386     87  1433     74    N     0

[145289 rows x 8 columns]

In [88]:
df.values[0][0]

1985

# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [ ]:
for index, row in df.iterrows():
    print row
    if index == 1:
        break

# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [ ]:
df[['Wscore', 'Lscore']].head()

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [ ]:
df.loc[:, ['Wscore', 'Lscore']].head()

Note the difference is the return types when you use brackets and when you use double brackets. 

In [ ]:
type(df['Wscore'])

In [ ]:
type(df[['Wscore']])

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [ ]:
df[0:3]

Here's an equivalent using iloc

In [ ]:
df.iloc[0:3,:]

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [ ]:
df.isnull().sum()

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Visualizing Data

An interesting way of displaying Dataframes is through matplotlib. 

In [ ]:
import matplotlib.pyplot as plt
%matplotlib inline

In [ ]:
ax = df['Wscore'].plot.hist(bins=20)
ax.set_xlabel('Points for Winning Team')

# Creating Kaggle Submission CSVs

This isn't directly Pandas related, but I assume that most people who use Pandas probably do a lot of Kaggle competitions as well. As you probably know, Kaggle competitions require you to create a CSV of your predictions. Here's some starter code that can help you create that csv file

In [ ]:
import numpy as np
import csv

results = [[0,10],[1,15],[2,20]]
results = pd.np.array(results)
print results

In [ ]:
firstRow = [['id', 'pred']]
with open("result.csv", "wb") as f:
    writer = csv.writer(f)
    writer.writerows(firstRow)
    writer.writerows(results)

The approach I described above deals more with python lists and numpy. If you want a purely Pandas based approach, take a look at this video: https://www.youtube.com/watch?v=ylRlGCtAtiE&list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y&index=22

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y

##### ASD

_ddhjd_